### DISCUSSION TITLE
How to deal with plugin templates?

### DISCUSSION THREAD

**Discussion by kilianso** [timber/timber#2623] [Q&A]
Hi all! I've seen some discussion on the topic of how to use Timber in plugin templates. But this is **not** my question. I use Learndash to sell courses. Learndash has a lot of templates and assets that are in the plugin directory in a theme called ld30. There are routes like `/courses/courseXY` and `/lessons/lessonXY`.

Now instead of picking the plugin templates, my Timber based theme just uses the default `archive.php` and `single.php` files when I go to these routes. And of course, because of this, the entire user interface and functionality of the plugin doesn't work. What am I missing?

Even if I could create pages like `archive-sfwd-courses.php` and `single-sfwd-courses.php` etc and then place plugin shortcodes and functions on them, it would be a huge effort and most likely I would miss some of the functions and edge cases covered by the built-in templates. So I would prefer to use their templates rather than create my own ones from scratch. Even if I copy all the related files to my theme's root directory (as [stated in this article](https://billosuch.com/2019/07/04/customizing-learndash-3-0-templates/)), it doesn't work.

If I switch to another Wordpress default theme that is not based on Timber, it works as expected. It's hard to determine if this is a compatibility issue between Learndash and Timber, or something general that I'm overlooking that occurs with any plugin that provides its own templates.

Here's the theme i'm building: https://github.com/kilianso/woodplane

**Comment by kilianso** [comment]
Okay, I looked into this a little more and came across a lot of discussions that are all related to version 2 of Timber.

Like these two:
https://github.com/timber/timber/issues/1639
https://github.com/timber/timber/issues/2501

I'm running on 1.20 and got it working on `single.php` pages like this:

```php
if (have_posts()) {
  while(have_posts()) {
    the_post(); 

    $post = new Timber\Post();
    $context['post'] = $post;

    Timber::render('single.twig', $context, 600);
  }
}
```

Is this a legit way of doing it? Are there other, better options or would you recommend to just migrate to version 2?

**Reply by gchtr** [reply]
Thanks for looking into this. Plugin compatibility sometimes is a pain. For premium plugins like Learndash it’s even a little harder, because they don’t seem to have a free version we could use and test. The issues you found could very well be what the underlying problem is. Could you check whether Learndash uses the `loop_start` or the `the_post` hook?

For Timber 1.x, your solution looks like a good way of doing it. When you use `have_posts()` and `the_posts()`, WordPress will call all revelant functions to set up the post object. Learndash possibly hooks into some of the filters in those functions.

In the long run, it’s probably better to migrate to Timber version 2. But for now, you should be fine. I think your way of doing it should even be fine for Timber v2. Or you could upgrade to the new `Post::setup()` function ([see Upgrade Guide](https://timber.github.io/docs/v2/upgrade-guides/2.0/#template-context)).

**Reply by kilianso** [reply]
Hey @gchtr 
Thanks for your answer and advice. Yes, as it seems they use hooks like `the_post`. Here in their [docs is an example.](https://developers.learndash.com/snippet/enroll-user-in-all-free-courses-upon-registration/)

So this solution should work for `single.php` and `page.php` right? But what about `archive.php` pages? How would you make them compatible if a plugin like Learndash has its own templates?

A simplified version of my current `archive.php` looks like this:

```php
$context = Timber::context();
$context['posts'] = Timber::get_posts();
Timber::render('archive.twig', $context, 600);
```

**Comment by langhenet** [comment]
Hi @kilianso 

Provided that I have "timberized" the header and footer of my theme,
I actually got learndash working by creating a `single-sfwd-course.php`

which inside has

```
<?php
  get_header();

  // Get the timber context
  $context = Timber::get_context();
  // Loads the Timber post class and sends it to the $context variable
  $timber_post = Timber::get_post();
  $context['post'] = $timber_post;
?>
</head>
<body>
<?php
  if ( have_posts() ) {

    while ( have_posts() ) {
      the_post();
      
      the_content();
    }
  }

  get_footer();
?>
  </body>
</html>
```

Edit: of course the markup part depends on what you have in your theme :-)
Edit 2: nvm, I just see that you have already solved it 😅

**Reply by kilianso** [reply]
Hey @langhenet
Thanks for you answer. I have a class based setup and my main Theme.php (where all the autoloading is going on) is already handling `get_header` and `get_footer` for all needed routes. This way, it's not necessary to do it again on specific page templates. However the rest is quite similar to the solution i posted above. Except your additional  `the_content()` call. What's up with that? I thought that's already covered by doing this:

```php
$post = new Timber\Post();
$context['post'] = $post;
```

One more thing to mention: I also tried your approach creating a `single-sfwd-course.php` but my goal is to make my theme compatible with any plugin, not just Learndash specifically.

**Reply by langhenet** [reply]
Hi @kilianso, yes also my theme has already the the get_footer/header calls, but I had to use this workaround to make it work :-) (so I'm extending a twig template which do not includes those calls.

Also, the the_content() call was necessary to have it working.

I was extending a parent theme with the need to only work with learndash though :-)
